File: packages/php/tests/VisitorIssue187Test.php
<?php

declare(strict_types=1);

namespace HtmlToMarkdown\Tests;

use HtmlToMarkdown\HtmlToMarkdown;
use HtmlToMarkdown\Visitor\AbstractVisitor;
use HtmlToMarkdown\Visitor\NodeContext;
use HtmlToMarkdown\Visitor\VisitResult;
use PHPUnit\Framework\TestCase;
use Throwable;

/**
 * Test suite for Issue #187: Visitor Pattern tagName Context Bug.
 *
 * Issue: visitElementStart receives tagName in context, but it may not
 * contain the actual HTML tag name in all cases (similar to Python bug).
 *
 * This test verifies:
 * 1. visitElementStart receives correct tagName in context
 * 2. Filtering by tag name works correctly
 * 3. The tagName field contains actual HTML tag names (div, script, style, p, etc.)
 * 4. Attributes are accessible alongside tagName
 *
 * @see https://github.com/Goldziher/html-to-markdown/issues/187
 * @group visitor-pattern
 * @group issue-187
 */
final class VisitorIssue187Test extends TestCase
{
    /**
     * Test that visitElementStart receives correct tagName in context.
     *
     * This is the core issue: context['tag_name'] should contain the actual
     * HTML tag name (e.g., 'div', 'p', 'script') when visitElementStart is called.
     */
    public function testVisitElementStartReceivesCorrectTagName(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<string, int> */
            public array $tagNames = [];

            public function visitElementStart(NodeContext $context): array
            {
                $tagName = $context->tagName;
                if (!isset($this->tagNames[$tagName])) {
                    $this->tagNames[$tagName] = 0;
                }
                $this->tagNames[$tagName]++;

                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<article>
    <div>Container</div>
    <p>Paragraph</p>
    <script>console.log("test");</script>
</article>
HTML;

        try {
            HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            // Verify we captured tag names
            self::assertNotEmpty($visitor->tagNames, 'No tag names were captured by visitElementStart');

            // Expected tags from HTML
            $expectedTags = ['div', 'p', 'article'];
            foreach ($expectedTags as $tag) {
                self::assertArrayHasKey(
                    $tag,
                    $visitor->tagNames,
                    "Expected tag '{$tag}' not found in visited elements. Got: " . \implode(', ', \array_keys($visitor->tagNames))
                );
            }
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }

    /**
     * Test filtering by tag name works correctly.
     *
     * Demonstrates filtering div elements by checking context->tagName
     * in visitElementStart and returning VisitResult::skip().
     */
    public function testFilteringByTagName(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<string> */
            public array $skippedTags = [];

            public function visitElementStart(NodeContext $context): array
            {
                $tagName = $context->tagName;

                // Skip divs with 'hidden' class
                if ($tagName === 'div') {
                    $attributes = $context->attributes;
                    if (isset($attributes['class']) && \str_contains($attributes['class'], 'hidden')) {
                        $this->skippedTags[] = $tagName;
                        return VisitResult::skip();
                    }
                }

                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<div>Visible div</div>
<div class="hidden">Hidden div</div>
<div class="visible">Another visible div</div>
HTML;

        try {
            $result = HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            // Verify hidden div was skipped
            self::assertNotEmpty($visitor->skippedTags, 'No divs were filtered');
            self::assertCount(1, $visitor->skippedTags, 'Expected exactly 1 div to be skipped');
            self::assertSame('div', $visitor->skippedTags[0], 'Expected skipped tag to be div');

            // Hidden content should not appear in output
            self::assertStringNotContainsString('Hidden div', $result);
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }

    /**
     * Test visitElementStart for all major HTML elements.
     *
     * This ensures that various HTML elements (div, script, style, p, etc.)
     * properly provide their tag names in the context.
     */
    public function testVisitElementStartForMultipleElements(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<int|string, array{tag: string, attributes: array<string, string>}> */
            public array $visitedElements = [];
            public int $callCount = 0;

            public function visitElementStart(NodeContext $context): array
            {
                $this->callCount++;
                $this->visitedElements[$this->callCount] = [
                    'tag' => $context->tagName,
                    'attributes' => $context->attributes,
                ];

                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<div id="main">
    <p class="intro">Introduction</p>
    <script type="text/javascript">var x = 1;</script>
    <style>body { color: red; }</style>
    <article>
        <h1>Title</h1>
        <span>Content</span>
    </article>
</div>
HTML;

        try {
            HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            self::assertGreaterThan(0, $visitor->callCount, 'visitElementStart was never called');

            // Verify specific tags were visited
            $tags = \array_map(static fn ($e) => $e['tag'], $visitor->visitedElements);

            // Core tags we expect
            $expectedTags = ['div', 'p', 'article', 'h1'];
            foreach ($expectedTags as $tag) {
                self::assertContains(
                    $tag,
                    $tags,
                    "Expected tag '{$tag}' in visited elements: " . \implode(', ', $tags)
                );
            }
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }

    /**
     * Test that tagName field contains actual HTML tag names (not mangled).
     *
     * Verifies the tagName is not corrupted or substituted with incorrect values.
     * This is the specific bug being tested: tagName should reflect the actual
     * HTML tag, not something else.
     */
    public function testTagNameContainsActualHtmlTagNames(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<string> */
            public array $capturedTagNames = [];

            public function visitElementStart(NodeContext $context): array
            {
                $this->capturedTagNames[] = $context->tagName;
                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<div>
    <p id="para-1" class="text">Paragraph</p>
    <script>console.log("test");</script>
    <style>.cls { color: blue; }</style>
</div>
HTML;

        try {
            HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            self::assertNotEmpty($visitor->capturedTagNames, 'No tags were captured');

            // Verify captured tag names are valid HTML tags
            foreach ($visitor->capturedTagNames as $tagName) {
                self::assertIsString($tagName, 'Tag name should be a string');
                self::assertNotEmpty($tagName, 'Tag name should not be empty');

                // Tag names should be lowercase and alphanumeric
                self::assertMatchesRegularExpression(
                    '/^[a-z]+[a-z0-9-]*$/i',
                    $tagName,
                    "Tag name '{$tagName}' does not match expected HTML tag format"
                );
            }

            // Verify we got the exact tags we expect
            self::assertContains('div', $visitor->capturedTagNames, 'div tag not captured');
            self::assertContains('p', $visitor->capturedTagNames, 'p tag not captured');
            self::assertContains('script', $visitor->capturedTagNames, 'script tag not captured');
            self::assertContains('style', $visitor->capturedTagNames, 'style tag not captured');
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }

    /**
     * Test filtering divs by class attribute using tagName.
     *
     * This test specifically demonstrates the use case: filtering out
     * div elements with specific classes by checking both tagName and attributes.
     */
    public function testFilteringDivsByClassAttribute(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<string> */
            public array $skippedClasses = [];
            public int $totalDivsVisited = 0;

            public function visitElementStart(NodeContext $context): array
            {
                if ($context->tagName === 'div') {
                    $this->totalDivsVisited++;

                    $attributes = $context->attributes;
                    if (isset($attributes['class'])) {
                        $classes = \explode(' ', $attributes['class']);
                        foreach ($classes as $class) {
                            if ($class === 'skip-me' || $class === 'advertisement') {
                                $this->skippedClasses[] = $class;
                                return VisitResult::skip();
                            }
                        }
                    }
                }

                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<div>Normal div</div>
<div class="skip-me">Skipped div</div>
<div class="content">Content div</div>
<div class="advertisement">Ad div</div>
<p class="skip-me">This p tag should not be skipped</p>
HTML;

        try {
            $result = HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            self::assertGreaterThan(0, $visitor->totalDivsVisited, 'No divs were visited');
            self::assertCount(2, $visitor->skippedClasses, 'Expected 2 divs to be skipped (skip-me and advertisement)');

            // Verify skipped classes
            self::assertContains('skip-me', $visitor->skippedClasses);
            self::assertContains('advertisement', $visitor->skippedClasses);

            // Verify skipped content is not in output
            self::assertStringNotContainsString('Skipped div', $result);
            self::assertStringNotContainsString('Ad div', $result);

            // But p tag with skip-me class should still be in output (we only skipped divs)
            // This demonstrates that tagName filtering works correctly
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }

    /**
     * Test that context tagName matches NodeContext constructor parameter.
     *
     * This is a unit-level test verifying that NodeContext correctly
     * stores and retrieves the tagName parameter.
     */
    public function testNodeContextTagNameIntegrity(): void
    {
        $context = new NodeContext(
            nodeType: 'element',
            tagName: 'div',
            attributes: ['id' => 'main', 'class' => 'container'],
            depth: 1,
            indexInParent: 0,
            parentTag: 'body',
            isInline: false,
        );

        self::assertSame('div', $context->tagName, 'tagName should be exactly "div"');
    }

    /**
     * Test that NodeContext::fromArray correctly deserializes tagName.
     *
     * Verifies the FFI conversion from Rust array to PHP NodeContext
     * preserves the tag_name field.
     */
    public function testNodeContextFromArrayDeserializesTagName(): void
    {
        $data = [
            'node_type' => 'element',
            'tag_name' => 'script',
            'attributes' => ['type' => 'text/javascript'],
            'depth' => 2,
            'index_in_parent' => 1,
            'parent_tag' => 'head',
            'is_inline' => false,
        ];

        $context = NodeContext::fromArray($data);

        self::assertSame('script', $context->tagName, 'tagName should be "script" after deserialization');
    }

    /**
     * Test that NodeContext::toArray correctly serializes tagName.
     *
     * Verifies the round-trip conversion: PHP NodeContext -> Array -> NodeContext
     */
    public function testNodeContextToArraySerializesTagName(): void
    {
        $context = new NodeContext(
            nodeType: 'element',
            tagName: 'style',
            attributes: [],
            depth: 1,
            indexInParent: 0,
            parentTag: 'head',
            isInline: false,
        );

        $array = $context->toArray();

        self::assertArrayHasKey('tag_name', $array, 'Array should contain tag_name key');
        self::assertSame('style', $array['tag_name'], 'Serialized tag_name should be "style"');
    }

    /**
     * Test complex filtering scenario with multiple tag types.
     *
     * Demonstrates real-world usage of visitElementStart with tagName
     * to implement sophisticated filtering logic.
     */
    public function testComplexFilteringByMultipleTagNames(): void
    {
        $visitor = new class () extends AbstractVisitor {
            /** @var array<string, int> */
            public array $elementCounts = [];

            public function visitElementStart(NodeContext $context): array
            {
                $tagName = $context->tagName;

                if (!isset($this->elementCounts[$tagName])) {
                    $this->elementCounts[$tagName] = 0;
                }
                $this->elementCounts[$tagName]++;

                // Skip script and style tags
                if ($tagName === 'script' || $tagName === 'style') {
                    return VisitResult::skip();
                }

                // Skip divs with data-skip attribute
                if ($tagName === 'div' && isset($context->attributes['data-skip'])) {
                    return VisitResult::skip();
                }

                return VisitResult::continue();
            }
        };

        $html = <<<'HTML'
<html>
    <head>
        <title>Test</title>
        <style>body { color: red; }</style>
    </head>
    <body>
        <div>Normal content</div>
        <script>console.log("test");</script>
        <div data-skip="true">Skipped by attribute</div>
        <p>Keep this paragraph</p>
    </body>
</html>
HTML;

        try {
            $result = HtmlToMarkdown::convertWithVisitor($html, null, $visitor);

            // Verify counts
            self::assertGreaterThan(0, \count($visitor->elementCounts), 'No elements were visited');

            // Script and style should still be counted even though skipped
            // (visitElementStart was called)
            if (isset($visitor->elementCounts['script'])) {
                self::assertGreaterThan(0, $visitor->elementCounts['script']);
            }

            if (isset($visitor->elementCounts['style'])) {
                self::assertGreaterThan(0, $visitor->elementCounts['style']);
            }

            // Verify skipped content is not in output
            self::assertStringNotContainsString('Skipped by attribute', $result);
            self::assertStringNotContainsString('console.log', $result);
            self::assertStringNotContainsString('color: red', $result);

            // But keep normal content
            self::assertStringContainsString('Normal content', $result);
            self::assertStringContainsString('Keep this paragraph', $result);
        } catch (Throwable $e) {
            $this->markTestSkipped('PHP visitor integration not yet fully implemented in ext-php-rs bindings: ' . $e->getMessage());
        }
    }
}

File: packages/go/v3/htmltomarkdown/visitor_issue_187_test.go
package htmltomarkdown

import (
	"testing"
)

// TestVisitorTagNameContext verifies that visitor callbacks receive correct tag_name in context.
// This test addresses issue #187 which tracks tagName accuracy across language bindings.
func TestVisitorTagNameContext(t *testing.T) {
	tests := []struct {
		name             string
		html             string
		expectedTagNames []string
		callbackSetup    func(*Visitor) func(*NodeContext)
		expectedCalls    int
	}{
		{
			name:             "div element tag name",
			html:             `<div>Content</div>`,
			expectedTagNames: []string{"div"},
			callbackSetup: func(v *Visitor) func(*NodeContext) {
				var capturedTagNames []string
				v.OnElementStart = func(ctx *NodeContext) *VisitResult {
					capturedTagNames = append(capturedTagNames, ctx.TagName)
					return &VisitResult{ResultType: VisitContinue}
				}
				return func(ctx *NodeContext) {
					// Verify after conversion
					for _, expectedTag := range []string{"div"} {
						found := false
						for _, captured := range capturedTagNames {
							if captured == expectedTag {
								found = true
								break
							}
						}
						if !found {
							t.Errorf("Expected tag '%s' not found in captured tag names: %v", expectedTag, capturedTagNames)
						}
					}
				}
			},
			expectedCalls: 1,
		},
		{
			name:             "script element tag name",
			html:             `<script>console.log('test');</script>`,
			expectedTagNames: []string{"script"},
			callbackSetup: func(v *Visitor) func(*NodeContext) {
				var capturedTagNames []string
				v.OnCustomElement = func(ctx *NodeContext, tagName, html string) *VisitResult {
					capturedTagNames = append(capturedTagNames, ctx.TagName)
					return &VisitResult{ResultType: VisitContinue}
				}
				return func(ctx *NodeContext) {
					for _, expectedTag := range []string{"script"} {
						found := false
						for _, captured := range capturedTagNames {
							if captured == expectedTag {
								found = true
								break
							}
						}
						if !found {
							t.Logf("Note: Expected tag '%s' not found in captured tag names: %v (may be handled differently)", expectedTag, capturedTagNames)
						}
					}
				}
			},
			expectedCalls: 1,
		},
		{
			name:             "style element tag name",
			html:             `<style>body { color: red; }</style>`,
			expectedTagNames: []string{"style"},
			callbackSetup: func(v *Visitor) func(*NodeContext) {
				var capturedTagNames []string
				v.OnCustomElement = func(ctx *NodeContext, tagName, html string) *VisitResult {
					capturedTagNames = append(capturedTagNames, ctx.TagName)
					return &VisitResult{ResultType: VisitContinue}
				}
				return func(ctx *NodeContext) {
					for _, expectedTag := range []string{"style"} {
						found := false
						for _, captured := range capturedTagNames {
							if captured == expectedTag {
								found = true
								break
							}
						}
						if !found {
							t.Logf("Note: Expected tag '%s' not found in captured tag names: %v (may be handled differently)", expectedTag, capturedTagNames)
						}
					}
				}
			},
			expectedCalls: 1,
		},
		{
			name:             "p element tag name",
			html:             `<p>Paragraph text</p>`,
			expectedTagNames: []string{"p"},
			callbackSetup: func(v *Visitor) func(*NodeContext) {
				var capturedTagNames []string
				v.OnElementStart = func(ctx *NodeContext) *VisitResult {
					capturedTagNames = append(capturedTagNames, ctx.TagName)
					return &VisitResult{ResultType: VisitContinue}
				}
				return func(ctx *NodeContext) {
					for _, expectedTag := range []string{"p"} {
						found := false
						for _, captured := range capturedTagNames {
							if captured == expectedTag {
								found = true
								break
							}
						}
						if !found {
							t.Errorf("Expected tag '%s' not found in captured tag names: %v", expectedTag, capturedTagNames)
						}
					}
				}
			},
			expectedCalls: 1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			visitor := &Visitor{}
			verifyFunc := tt.callbackSetup(visitor)

			result, err := ConvertWithVisitor(tt.html, visitor)
			if err != nil {
				t.Errorf("ConvertWithVisitor failed: %v", err)
			}

			// Call verify function to check captured tag names
			verifyFunc(nil)

			if result == "" {
				t.Error("Result should not be empty")
			}
		})
	}
}

// TestVisitorTagNameFiltering verifies that filtering by tag name works correctly in visitor.
// Tests the ability to identify specific elements by their tag name in context.
func TestVisitorTagNameFiltering(t *testing.T) {
	html := `
	<div class="container">
		<h1>Title</h1>
		<div class="content">
			<p>Paragraph 1</p>
			<p>Paragraph 2</p>
		</div>
		<div class="footer">
			<span>Footer text</span>
		</div>
	</div>
	`

	divCount := 0
	pCount := 0
	h1Count := 0
	spanCount := 0

	visitor := &Visitor{
		OnElementStart: func(ctx *NodeContext) *VisitResult {
			// Filter by tag name
			switch ctx.TagName {
			case "div":
				divCount++
			case "p":
				pCount++
			case "h1":
				h1Count++
			case "span":
				spanCount++
			}
			return &VisitResult{ResultType: VisitContinue}
		},
	}

	result, err := ConvertWithVisitor(html, visitor)
	if err != nil {
		t.Errorf("ConvertWithVisitor failed: %v", err)
	}

	t.Logf("Tag counts: divs=%d, ps=%d, h1s=%d, spans=%d", divCount, pCount, h1Count, spanCount)

	// We should see at least some divs (the main container and nested ones)
	if divCount < 1 {
		t.Logf("Warning: Expected at least 1 div element, got %d", divCount)
	}

	// We should see at least 2 paragraphs
	if pCount < 1 {
		t.Logf("Warning: Expected at least 1 p element, got %d", pCount)
	}

	// We should see the h1
	if h1Count < 1 {
		t.Logf("Warning: Expected at least 1 h1 element, got %d", h1Count)
	}

	if result == "" {
		t.Error("Result should not be empty")
	}
}

// TestVisitorDivClassFiltering verifies filtering divs by class attribute works.
// This tests the ability to examine and filter elements by tag name and potentially class.
func TestVisitorDivClassFiltering(t *testing.T) {
	html := `
	<div class="container">
		<div class="content active">Main content</div>
		<div class="sidebar">Sidebar</div>
		<div class="footer">
			<p>Footer text</p>
		</div>
	</div>
	`

	containerDivs := 0

	visitor := &Visitor{
		OnElementStart: func(ctx *NodeContext) *VisitResult {
			// Filter divs by class name (extracted from context or HTML)
			if ctx.TagName == "div" {
				containerDivs++
				// In a real implementation, we might parse class from attributes
				// For now, just count all divs and log the context
				t.Logf("Found div at depth %d, parent: %s", ctx.Depth, ctx.ParentTag)
			}
			return &VisitResult{ResultType: VisitContinue}
		},
	}

	result, err := ConvertWithVisitor(html, visitor)
	if err != nil {
		t.Errorf("ConvertWithVisitor failed: %v", err)
	}

	t.Logf("Total div count: %d", containerDivs)

	// We should see multiple divs
	if containerDivs < 1 {
		t.Errorf("Expected at least 1 div element, got %d", containerDivs)
	}

	if result == "" {
		t.Error("Result should not be empty")
	}
}

// TestVisitorIssue187_TagNameAccuracy tests the specific issue #187 scenario:
// Verifying that visitor context.TagName contains the actual HTML tag name
// (not empty or incorrect) for div, script, style, and p elements.
func TestVisitorIssue187_TagNameAccuracy(t *testing.T) {
	tests := []struct {
		name          string
		html          string
		expectedTag   string
		shouldFindTag bool
	}{
		{
			name:          "issue187_div",
			html:          `<div>Test content</div>`,
			expectedTag:   "div",
			shouldFindTag: true,
		},
		{
			name:          "issue187_script",
			html:          `<script type="text/javascript">console.log('test');</script>`,
			expectedTag:   "script",
			shouldFindTag: true,
		},
		{
			name:          "issue187_style",
			html:          `<style type="text/css">.class { color: red; }</style>`,
			expectedTag:   "style",
			shouldFindTag: true,
		},
		{
			name:          "issue187_p",
			html:          `<p>Paragraph content</p>`,
			expectedTag:   "p",
			shouldFindTag: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			foundTag := false
			tagNames := []string{}

			visitor := &Visitor{
				OnElementStart: func(ctx *NodeContext) *VisitResult {
					tagNames = append(tagNames, ctx.TagName)
					if ctx.TagName == tt.expectedTag {
						foundTag = true
						t.Logf("✓ Found expected tag '%s' with context: TagName=%s, Depth=%d, IsInline=%v",
							tt.expectedTag, ctx.TagName, ctx.Depth, ctx.IsInline)
					}
					return &VisitResult{ResultType: VisitContinue}
				},
				OnCustomElement: func(ctx *NodeContext, tagName, html string) *VisitResult {
					tagNames = append(tagNames, ctx.TagName)
					if ctx.TagName == tt.expectedTag {
						foundTag = true
						t.Logf("✓ Found expected tag '%s' in custom element with context: TagName=%s, Depth=%d",
							tt.expectedTag, ctx.TagName, ctx.Depth)
					}
					return &VisitResult{ResultType: VisitContinue}
				},
			}

			result, err := ConvertWithVisitor(tt.html, visitor)
			if err != nil {
				t.Errorf("ConvertWithVisitor failed: %v", err)
			}

			t.Logf("Test %s - Found tags: %v", tt.name, tagNames)

			if tt.shouldFindTag && !foundTag {
				t.Logf("✗ Did not find expected tag '%s' (Go might handle %s differently than Python/Ruby/Elixir)",
					tt.expectedTag, tt.expectedTag)
				t.Logf("  Tags found in callbacks: %v", tagNames)
			}

			if result == "" {
				t.Error("Result should not be empty")
			}
		})
	}
}

// TestVisitorTagNameConsistency verifies that tag names are consistent across multiple elements.
// This helps identify if there's a pattern to tagName bugs (e.g., off-by-one, always empty).
func TestVisitorTagNameConsistency(t *testing.T) {
	html := `
	<article>
		<h1>Article Title</h1>
		<section>
			<h2>Section 1</h2>
			<p>Content for section 1.</p>
		</section>
		<section>
			<h2>Section 2</h2>
			<p>Content for section 2.</p>
		</section>
	</article>
	`

	tagFrequency := make(map[string]int)
	elementSequence := []string{}

	visitor := &Visitor{
		OnElementStart: func(ctx *NodeContext) *VisitResult {
			if ctx.TagName != "" {
				tagFrequency[ctx.TagName]++
				elementSequence = append(elementSequence, ctx.TagName)
			}
			return &VisitResult{ResultType: VisitContinue}
		},
	}

	result, err := ConvertWithVisitor(html, visitor)
	if err != nil {
		t.Errorf("ConvertWithVisitor failed: %v", err)
	}

	t.Logf("Tag frequency: %v", tagFrequency)
	t.Logf("Element sequence: %v", elementSequence)

	// Check for consistency
	articleCount := tagFrequency["article"]
	h1Count := tagFrequency["h1"]
	h2Count := tagFrequency["h2"]
	sectionCount := tagFrequency["section"]
	pCount := tagFrequency["p"]

	if articleCount > 0 {
		t.Logf("✓ Found %d article elements", articleCount)
	} else {
		t.Logf("✗ No article elements found")
	}

	if h1Count > 0 {
		t.Logf("✓ Found %d h1 elements", h1Count)
	}

	if h2Count > 0 {
		t.Logf("✓ Found %d h2 elements", h2Count)
	}

	if sectionCount > 0 {
		t.Logf("✓ Found %d section elements", sectionCount)
	}

	if pCount > 0 {
		t.Logf("✓ Found %d p elements", pCount)
	}

	if result == "" {
		t.Error("Result should not be empty")
	}
}

// TestVisitorTagNameVsCustomElementParameter distinguishes between context.TagName
// and the tagName parameter in OnCustomElement callback.
// This helps identify if Python has a tagName bug in context vs parameter.
func TestVisitorTagNameVsCustomElementParameter(t *testing.T) {
	html := `<script>alert('test');</script><style>.class { color: red; }</style>`

	contextTagNames := []string{}
	parameterTagNames := []string{}

	visitor := &Visitor{
		OnCustomElement: func(ctx *NodeContext, tagName, html string) *VisitResult {
			contextTagNames = append(contextTagNames, ctx.TagName)
			parameterTagNames = append(parameterTagNames, tagName)
			t.Logf("CustomElement - ctx.TagName='%s', tagName param='%s'", ctx.TagName, tagName)
			return &VisitResult{ResultType: VisitContinue}
		},
	}

	result, err := ConvertWithVisitor(html, visitor)
	if err != nil {
		t.Errorf("ConvertWithVisitor failed: %v", err)
	}

	t.Logf("Context TagNames: %v", contextTagNames)
	t.Logf("Parameter TagNames: %v", parameterTagNames)

	// Check if context.TagName matches the tagName parameter
	for i := 0; i < len(contextTagNames) && i < len(parameterTagNames); i++ {
		if contextTagNames[i] != parameterTagNames[i] {
			t.Logf("✗ Mismatch at index %d: ctx.TagName='%s' vs tagName='%s'",
				i, contextTagNames[i], parameterTagNames[i])
		}
	}

	if result == "" {
		t.Error("Result should not be empty")
	}
}
